### ISSUE TITLE
ShortPrefixPassed should still allow 2 character prefixes if desired.

### ISSUE THREAD

**Issue by aLANparty** [WordPress/WordPress-Coding-Standards#1836]
## Bug Description
The addition of the ShortPrefixPassed error breaks backwards compatibility with 2 character prefixes, but doesn't provide any way to restore support for them other than totally disabling the entire PrefixAllGlobals sniff.

The ShortPrefixPassed error itself can be excluded, but because the code which triggers it prevents the prefix from ever being added to the list of configured prefixes it break other sniffs even when it is excluded. 

I believe the error should stay as is, but the "continue;" line following it, which prevents adding the prefixes, should be removed to restore support for short prefixes if desired.  Another alternative would be to allow the user to override the minimum prefix length.

## Minimal Code Snippet
```xml
<rule ref="WordPress">
	<exclude name="WordPress.NamingConventions.PrefixAllGlobals.ShortPrefixPassed"/>
</rule>
<rule ref="WordPress.NamingConventions.PrefixAllGlobals">
    <properties>
        <property name="prefixes" type="array">
            <element value="xx"/>
        </property>
    </properties>
</rule>
```

## Error Code
WordPress.NamingConventions.PrefixAllGlobals.ShortPrefixPassed

## Environment

| Question               | Answer
| ------------------------| -------
| PHP version             | 7.2
| PHP_CodeSniffer version | 3.5.2
| WPCS version            | 2.2.0
| WPCS install type       | Composer project local
| IDE (if relevant)       | vscode

## Tested Against `develop` branch?
- [ x] I have verified the issue still exists in the `develop` branch of WPCS.

**Comment by jrfnl** [comment]
@aLANparty Let's do a simple calculation: there are over 80.000 plugins + more than 5.000 themes available via wp.org. That's not even taking the multitudes of commercial plugins and themes into account.

The prefix for each plugin/theme should preferably be unique as otherwise you get into conflict territory.

Now with a two character prefix, that means 27 (a-z_) * 37 (a-z0-9_) = 999 unique possibilities...

So, there is a reason that rule is now being enforced. Two character prefixes is just a _very bad idea_. Even three character prefixes is stretching it.

@Ipstenu Would you like to chime in ?

**Comment by aLANparty** [comment]
I understand the reason for the error existing.  I'm certainly not arguing that it isn't a good idea. The issue is just that phpcs is generally used in a manner that allows the user to customize the standard to their particular needs.  It doesn't seem reasonable to require anyone who uses the standard, but has legacy code with 2 letter prefixes, to change all their code.  People will just disable the entire rule which is worse.

**Comment by jrfnl** [comment]
> It doesn't seem reasonable to require anyone who uses the standard, but has legacy code with 2 letter prefixes, to change all their code.

Well, considering support for PHP < 5.6 has been dropped now, wouldn't now be a good time to fix it properly and change the code base over to use namespaces ?

**Comment by jrfnl** [comment]
>The issue is just that phpcs is generally used in a manner that allows the user to customize the standard to their particular needs.

While that is true for most code style related things, this sniff is not about code style.

Conflicting prefixes/function names/class names will cause real life white-screens of death for users, so isn't really up for discussion or personal preference.

**Comment by aLANparty** [comment]
My point is that the way PrefixAllGlobals.ShortPrefixPassed is implemented means that in order to disable it requires also disabling PrefixAllGlobals.  I don't understand how that is beneficial to anyone.   The only result is that anyone who wants to allow 2 letter prefixes will loose the ability to check for prefixes entirely.

**Comment by jrfnl** [comment]
Well, checking for two character prefixes is next to useless, so yes, in that case, there's no point in running the sniff.

**Comment by aLANparty** [comment]
PrefixAllGlobals can be used to check for more than one prefix.  So you don't just loose the ability to check for just 2 character prefixes.  Any project with multiple prefixes which happens to have at least one legacy 2 character prefix losses the ability to check for all of them.

**Comment by jrfnl** [comment]
@aLANparty The sniff will still check against the other prefixes, it will just ignore the two-char prefix and start throwing errors for everything using that.

**Comment by aLANparty** [comment]
That is the issue.

**Comment by aLANparty** [comment]
Sorry for being curt in my last reply.  What I mean is the issue is I now have dozens of projects with a lint error and the only way to fix it is either disable PrefixAllGlobals or refactor all the code.  The first is not ideal, but the second is nonstarter which means disabling PrefixAllGlobals is my only choice.  The change I'm proposing would enable me to keep PrefixAllGlobals active and wouldn't cause any loss of functionality for others.

**Comment by jrfnl** [comment]
Well, you could revert back to using WPCS 2.1.1 for the time being until you've fixed the prefixes in the problem projects...

**Comment by aLANparty** [comment]
If this were a 3.0 release I would understand this view point, but the needless loss of backwards compatibility for a point release seems unnessecary.  Most people are going to be using the standard with composer and semver is going to cause them to upgrade to this version.  They will then get PrefixAllGlobals.ShortPrefixPassed errors (as expected) but also start getting PrefixAllGlobals.NonPrefixed*Found errors all of a sudden.  And you could argue that the PrefixAllGlobals.NonPrefixed*Found aren't even correct.  

What is the advantage of having a two letter prefix trigger both PrefixAllGlobals.ShortPrefixPassed and PrefixAllGlobals.NonPrefixed*Found anyway? Surely PrefixAllGlobals.ShortPrefixPassed is sufficient to explain the issue.

**Comment by Ipstenu** [comment]
Option 3 would be to edit the code and not check that check ON those functions only. It's a hassle, but it'll also prep you for fixing your problems properly in the long run.

But yes, this is a huge issue with plugins. No prefixes and two letter prefixes are an everyday occurrence. It's one of the most common reasons a plugin is held up for review.

**Comment by aLANparty** [comment]
Oh well. <exclude name="WordPress.NamingConventions.PrefixAllGlobals"/> it is then.  It's a shame.

**Comment by GaryJones** [comment]
> Oh well. `<exclude name="WordPress.NamingConventions.PrefixAllGlobals"/>` it is then. It's a shame.

I must be missing something - why can't `<exclude name="WordPress.NamingConventions.PrefixAllGlobals.ShortPrefixPassed"/>` be used to just exclude that violation code?

**Comment by aLANparty** [comment]
In PrefixAllGlobalsSniff.php the section which triggers ShortPrefixPassed is in the function `validate_prefixes()`.  This function builds the array of valid whitelisted prefixes that can be passed in via the `prefixes` propery to PrefixAllGlobals 

ie
```
<rule ref="WordPress.NamingConventions.PrefixAllGlobals">
    <properties>
        <property name="prefixes" type="array">
            <element value="xx"/>
            <element value="other_prefix"/>
            <element value="third_prefix"/>
        </property>
    </properties>
</rule>
```

The problem is that the if statement in which the error ShortPrefixPassed is set ends with `continue;`.  This continue causes the loop to skip adding any two character whitelisted prefix specified in the property above from being added to the list.  The result is that any other PrefixAllGlobals related rules which use that whitelist property `$this->validated_prefixes` or method `is_prefixed()` will ignore that the prefix is explicitly whitelisted.  From what I can tell this applies to all NonPrefixedXXXFound rules (NonPrefixedFunctionFound, NonPrefixedClassFound, etc).

The result, and my issue, is that although the rule ShortPrefixPassed itself can be explicitly excluded any NonPrefixedXXXFound type rule will always spit out an error because it never knows that it was told to treat the specified two character prefix as valid.


The relevent code from PrefixAllGlobalsSniff.php is below.

line 985
```
if ( $prefix_length < self::MIN_PREFIX_LENGTH ) {
	$this->phpcsFile->addError(
		'The "%s" prefix is too short. Short prefixes are not unique enough and may cause name collisions with other code.',
		0,
		'ShortPrefixPassed',
		array( $prefix )
	);
	continue;
}
```

line 1005
```
// Lowercase the prefix to allow for direct compare.
$prefixes[ $key ] = $prefixLC;
```

**Comment by dingo-d** [comment]
Or, alternatively, you can just create a custom ruleset and copy the `PrefixAllGlobals` sniff, add the check for 2/3 character prefix, exclude the core one and use your own until you fix the old codebase 🤷‍♂

**Comment by aLANparty** [comment]
Yes there are many pointless and complex alternatives I could do. I guess my only real problem is why? What is the problem with the fix I'm proposing. I don't understand why anyone is opposed to it. 

It is a very simple one line fix which restores broken functionality without breaking the new functionality. 

I just don't understand and no one has offered up any reason for why it is a good idea other than what amounts to 'no you'.

**Comment by jrfnl** [comment]
@aLANparty Good reasons have been given, but none you want to hear.

Allowing two-char prefixes, even though we throw an error about them:
1. Leaves the sniff broken as having a two char prefix is just as bad a not having a prefix at all.
    The new check is a **fix** to the principle of the sniff, making it stricter. It is not new functionality.
2. Still doesn't prevent end-users from ending up with white screens of death because of conflicting names in plugins and themes.
3. Inconsistent.

As @Ipstenu said, you could _temporarily_ whitelist the offending code in-line with `// phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals` comments to create an action list for yourself to do later, or you can fix it properly, as you should.

This repo is about applying proper coding standards. The fact that you don't want to, does not mean we should make the checks worse for everyone else.

**Comment by aLANparty** [comment]
Alright. Do whatever you want. It's not my code. I've already excluded `PrefixAllGlobals` from all my projects so it doesn't really matter to me anymore.

**Comment by chyvak1831** [comment]
My theme has short naming 'df'.
is really strange that I can't use this as prefix.

**Comment by chyvak1831** [comment]
okay, I have now rename all prefixes from `df` into `dsf` to pass standards.

> Leaves the sniff broken as having a two char prefix is just as bad a not having a prefix at all.

I'm curious why 3 letter naming `dsf` is better then two-letter `df`? What the goal achieved by this?

**Comment by jrfnl** [comment]
@chyvak1831 There are thousands of plugins and themes. Changes of another plugin/theme using the same non-unique name as your theme are therefor high when only a two-letter prefix is used - believe me, people's fantasy for function/class/variable names is limited.

When naming clashes occur this can lead to white screens of death for end-users and/or extremely hard to debug problems for plugin/theme owners.

With a 2-letter prefix, there are only 936 unique prefixes available. With a three letter prefix, this goes up to 33.696, which greatly reduces the changes of naming conflicts.
